### PR TITLE
Replace process:exit in functions with Result + error handling

### PR DESCRIPTION
**Pull request by nicmr**
# Description

Hi Rami, I hope you've had a good start into 2024 :) I'm back with another PR on error handling.

- Currently, some methods simply call `std::process::exit` when encountering an error, which can cause problems
  1. It immediately exits the program and does not call any destructors (currently probably not a problem, but might in the future) like a panic would
  2. More importantly, it makes it hard to tell if a given function will cause the program to terminate, e.g. just reading the signature of a function `pub fn selectable_list(input: Vec<String>) -> String` does not tell you if it can cause the program to terminate, which can be very surprising to developers and requires you to always look at the implementation of a function.
- They also immediately print an error message, which makes it hard to wrap the error message with additional information, for example why the function was called

I'm looking forward to your feedback :)

Fixes # -

## Type of change

- Add thiserror crate to make the error types less verbose
- Add `src/error.rs` module for error types
- Fallible functions in `src/modes.rs` now return `Result<(), Error>` (or `Option<T>` where appropriate)
- Error messages are now defined on the error type
- Errors are now wrapped with the context or have the error context as a field where appropriate

### Changes from a user perspective:

Only error messages have changed.
- Failing to make a selection with skim
  - old: `No item selected`
  - new (context): `error: no item selected when prompted to select context`
  - new (namespace): `error: no item selected when prompted to select namespace`
- Trying to select a context that doesn't exist (`kc foo`)
  - old: `error: no context exists with the name: foo`
  - new: `error: failed to set context: no context exists with the name foo`

# How Has This Been Tested?

- [x] Manual end-to-end testing

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Ramilito** [comment]
Hey Nico! Super hectic start but hoping it will settle down soon 😓!
I take a look this weekend, super sorry for the delay 🙏

File: src/main.rs [hunk 1]
```diff
@@ -68,22 +71,32 @@ enum Mode {
 }
 
 impl Mode {
-    fn invoke(&self) {
+    fn invoke(&self) -> Result <(), Error> {
+        let args = Cli::parse();
```
Comment (Ramilito) [inline src/main.rs:75]: Nice!

File: src/modes.rs [hunk 2]
```diff
@@ -1,81 +1,73 @@
-use std::process;
+use crate::{commands::{self}, config, error::Error, Cli, DEST, KUBECONFIG};
 
-use crate::{commands::{self}, config, Cli, DEST, KUBECONFIG};
-
-fn selection(value: Option<String>, callback: fn() -> String) -> String {
```
Comment (Ramilito) [inline src/modes.rs:5]: Thanking you for resolving this ugly hack 👍

**Review by Ramilito — APPROVED** [review]

**Comment by Ramilito** [comment]
> More importantly, it makes it hard to tell if a given function will cause the program to terminate, e.g. just reading the signature of a function pub fn selectable_list(input: Vec<String>) -> String does not tell you if it can cause the program to terminate, which can be very surprising to developers and requires you to always look at the implementation of a function.

This is such a good point! Thank you for pointing it out!

**Comment by nicmr** [comment]
> Hey Nico! Super hectic start but hoping it will settle down soon 😓!

Keeping my fingers crossed for you! It's been quite hectic over here as well :cold_sweat:

**Comment by nicmr** [comment]
Should I try and fix the issue with the CI before merging? I don't think it's related to my changes specifically

**Comment by Ramilito** [comment]
I don't think so either, Il create a issue to make forks not need the phony secret from this repo (I'm guessing that it's the issue)

**Pull request merged**